biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java

```diff
@@ -0,0 +1,145 @@
+import java.util.regex.Pattern;
+
+import aQute.bnd.header.Attrs;
+import aQute.bnd.osgi.Descriptors.PackageRef;
+
+public class PermissionGenerator {
+	public static final String	KEY					= "permissions";
+
+	public static final String	PARAM_ALL			= "all";
+	public static final String	PARAM_ADMIN			= "admin";
+	public static final String	PARAM_CAPABILITY	= "capability";
+	public static final String	PARAM_PACKAGES		= "packages";
+	public static final String	PARAM_SERVICES		= "services";
+
+	private final Builder		builder;
+	private final List<String>	args;
+	private final StringBuilder	sb;
+
+	public PermissionGenerator(Builder builder, final String... args) {
+		this.builder = builder;
+
+		if (args.length == 2 && PARAM_ALL.equals(args[1])) {
+			this.args = Arrays.asList(PARAM_ADMIN, PARAM_CAPABILITY, PARAM_PACKAGES, PARAM_SERVICES);
+		} else {
+			this.args = Arrays.asList(args);
+		}
+
+		sb = new StringBuilder();
+	}
+
+	public String generate() {
+		sb.setLength(0);
+
+		for (String arg : args) {
+			if (PARAM_ADMIN.equals(arg)) {
+				generateAdmin();
+			} else if (PARAM_CAPABILITY.equals(arg)) {
+				generateCapability();
+			} else if (PARAM_PACKAGES.equals(arg)) {
+				generatePackages();
+			} else if (PARAM_SERVICES.equals(arg)) {
+				generateServices();
+			} else if (KEY.equals(arg))
+				;
+			else
+				builder.error("Invalid option in ${permissions}: %s", arg);
+		}
+
+		return sb.toString();
+	}
+
+	private void generateAdmin() {
+		sb.append("(org.osgi.framework.AdminPermission)");
+	}
+
+	private void generateCapability() {
+		for (String namespace : builder.getProvideCapability().keySet()) {
+			if (!Processor.isDuplicate(namespace)) {
+				sb.append("(org.osgi.framework.CapabilityPermission \"").append(namespace)
+						.append("\" \"provide\")\r\n");
+			}
+		}
+		for (String namespace : builder.getRequireCapability().keySet()) {
+			if (!Processor.isDuplicate(namespace)) {
+				sb.append("(org.osgi.framework.CapabilityPermission \"").append(namespace)
+						.append("\" \"require\")\r\n");
+			}
+		}
+	}
+
+	private void generatePackages() {
+		if (builder.getImports() != null) {
+			for (PackageRef imp : builder.getImports().keySet()) {
+				if (!imp.isJava()) {
+					sb.append("(org.osgi.framework.PackagePermission \"");
+					sb.append(imp);
+					sb.append("\" \"import\")\r\n");
+				}
+			}
+		}
+		if (builder.getExports() != null) {
+			for (PackageRef exp : builder.getExports().keySet()) {
+				sb.append("(org.osgi.framework.PackagePermission \"");
+				sb.append(exp);
+				sb.append("\" \"export\")\r\n");
+			}
+		}
+	}
+
+	private void generateServices() {
+		for (String declaredService : getDeclaredServices()) {
+			sb.append("(org.osgi.framework.ServicePermission \"").append(declaredService)
+					.append("\" \"register\")\r\n");
+		}
+		for (String referencedService : getReferencedServices()) {
+			sb.append("(org.osgi.framework.ServicePermission \"").append(referencedService).append("\" \"get\")\r\n");
```

Comment:
End with `\n`. We shouldn't use `\r\n` and we want a consistent output regardless of whether the bundle is built on *ix or windows.